org.eclipse.jdt.ui/core refactoring/org/eclipse/jdt/internal/corext/refactoring/util/ChangedValueChecker.java

```diff
@@ -0,0 +1,582 @@
+					parser.setKind(ASTParser.K_COMPILATION_UNIT);
+					parser.setSource(icu);
+					parser.setResolveBindings(true);
+					CompilationUnit compilationUnit= (CompilationUnit) parser.createAST(null);
+					final ASTNode perform= NodeFinder.perform(compilationUnit, tmp.getSourceRange());
+					if (perform instanceof MethodDeclaration && ((MethodDeclaration) perform).resolveBinding() != null) {
+						MethodDeclaration md= (MethodDeclaration) perform;
+						if (Modifier.isAbstract(md.resolveBinding().getModifiers()))
+							continue;
+						return md;
+					} else {
+						return null;
+					}
+				}
+			}
+			return null;
+		} catch (JavaModelException e) {
+			;
+		}
+		return null;
+	}
+
+	private static void findTypes(IType it, ITypeHierarchy ith, ArrayList<IType> iTypes) {
+		iTypes.add(it);
+
+		for (IType i : ith.getAllSubtypes(it)) {
+			iTypes.add(i);
+		}
+		return;
+	}
+
+	private ASTNode getOriginalExpression(ASTNode node) {
+		while (node != null) {
+			if (node instanceof ParenthesizedExpression) {
+				ParenthesizedExpression pe= (ParenthesizedExpression) node;
+				node= pe.getExpression();
+			} else if (node instanceof CastExpression) {
+				CastExpression ce= (CastExpression) node;
+				node= ce.getExpression();
+			} else
+				break;
+		}
+		return node;
+	}
+
+	class Elem {
+		Object memberKey;
+
+		Elem e;
+
+		public Elem(ASTNode node, boolean flag) {
+			super();
+			if (node instanceof SimpleName) {
+				SimpleName sn= (SimpleName) node;
+				IBinding resolveBinding= sn.resolveBinding();
+				if (resolveBinding != null && resolveBinding.getJavaElement() != null) {
+					if (resolveBinding.getJavaElement().getElementType() != IJavaElement.LOCAL_VARIABLE
+							|| flag)
+						this.memberKey= resolveBinding.getKey();
+				}
+			} else if (node instanceof QualifiedName) {
+				QualifiedName qn= (QualifiedName) node;
+				SimpleName sn= qn.getName();
+				this.memberKey= sn.resolveBinding() != null ? sn.resolveBinding().getKey() : null;
+				if (qn.resolveBinding() != null && qn.resolveBinding().getModifiers() != Modifier.STATIC)
+					this.e= new Elem(qn.getQualifier(), flag);
+			} else if (node instanceof FieldAccess) {
+				FieldAccess fa= (FieldAccess) node;
+				SimpleName sn= fa.getName();
+				this.memberKey= sn.resolveBinding() != null ? sn.resolveBinding().getKey() : null;
+				ASTNode expr= getOriginalExpression(fa.getExpression());
+				if (expr instanceof MethodInvocation) {
+					if (flag) {
+						this.e= new Elem((MethodInvocation) expr);
+					} else {
+						this.memberKey= null;
+					}
+				} else
+					this.e= new Elem(expr, flag);
+			} else if (node instanceof MethodInvocation) {
+				MethodInvocation mi= (MethodInvocation) node;
+				if (flag) {
+					this.e= new Elem(mi);
+				} else {
+					this.memberKey= null;
+				}
+			}
+		}
+
+		public Elem(MethodInvocation expr) { // use String to represent the instances of MethodInvocation
+			super();
+			this.memberKey= expr.toString();
```

Comment:
Do you need to use getOriginalExpression() for this?

Reply:
It has been ensured that `expr` will not be a ParenthesizedExpression before calling.
```java
ASTNode expr= getOriginalExpression(fa.getExpression());
if (expr instanceof MethodInvocation) {
	if (flag) {
		this.e= new Elem((MethodInvocation) expr);
...
```

Reply:
Ok.